File: matrix-chain-multiplication/striver_top_down_dp.cpp
#include <bits/stdc++.h> 

// Time complexity: O(n^3), Space complexity: O(n^2) + Auxiliary Stack Space: O(n)

int compMultns(int st, int end, vector<int>& arr, vector<vector<int>>& dp) {
    if (st == end) {
        return 0;
    }
    if (dp[st][end] != -1) {
        return dp[st][end];
    }

    int minMults = 1e9;

    for (int k = st; k < end; k++) {
        int mults = arr[st - 1]*arr[k]*arr[end] + compMultns(st, k,arr, dp) + compMultns(k + 1, end, arr, dp);
        minMults = min(minMults, mults);
    }

    return dp[st][end] = minMults;

}

int matrixMultiplication(vector<int> &arr, int N) {
    // Write your code here.
    vector<vector<int>> dp(N, vector<int>(N, -1));
    return compMultns(1, N - 1, arr, dp);
}

File: matrix_chain_multiplication/py_solns/striver_top_down_dp.py
# TC: O(n^3)
# SC: O(n*2) + Auxiliary stack space O(n)

# GeeksforGeeks link: https://www.geeksforgeeks.org/problems/matrix-chain-multiplication0303/1

class Solution:
    
    def init_dp_arr(self, rows, cols, init_val):
        dp = [[init_val for _ in range(cols)] for _ in range(rows)]
        return dp
    
    def comp_mc_operns(self, arr, i, j, dp):
        
        # Base case:
        if i == j:
            return 0
        if dp[i][j] != -1:
            return dp[i][j]
        min_steps = 1e9
        for k in range(i, j):
                    # num_steps to multiply the final matrix
                    # + num_steps to multiply the 2 sub-matrices separately.
            steps = arr[i-1]*arr[k]*arr[j] + self.comp_mc_operns(arr, i, k, dp) + self.comp_mc_operns(arr, k + 1, j, dp)
            
            min_steps = min(min_steps, steps)
        dp[i][j] = min_steps
        return min_steps
    
    def matrixMultiplication(self, arr):
        # code here
        
        n = len(arr)
        dp = self.init_dp_arr(n, n, -1)
        return self.comp_mc_operns(arr, 1, n - 1, dp)